test/bigquery.ts

```diff
@@ -708,26 +735,207 @@ describe('BigQuery', () => {
+      const type = BigQuery.getTypeDescriptorFromProvidedType_({prop: 'INT64'});
+
+      assert.deepStrictEqual(type, {
+        type: 'STRUCT',
+        structTypes: [
+          {
+            name: 'prop',
+            type: {
+              type: 'INT64',
+            },
+          },
+        ],
+      });
+    });
+
+    it('should throw for invalid provided types with named params', done => {
+      const INVALID_TYPES = {key: 'invalid'};
+
+      assert.throws(() => {
+        bq.createQueryJob({
+          query: QUERY_STRING,
+          params: NAMED_PARAMS,
+          types: INVALID_TYPES,
+        });
+      }, /Invalid type provided./);
+      done();
+    });
+
+    it('should throw if named param not present in provided types', done => {
+      const INVALID_TYPES = {other: 'string'};
+
+      assert.throws(() => {
+        bq.createQueryJob({
+          query: QUERY_STRING,
+          params: NAMED_PARAMS,
+          types: INVALID_TYPES,
+        });
+      }, /Type not provided for parameter: key/);
+      done();
+    });
+
+    it('should throw for invalid type structure provided for named params', done => {
+      assert.throws(() => {
+        bq.createQueryJob({
+          query: QUERY_STRING,
+          params: NAMED_PARAMS,
+          types: POSITIONAL_PARAMS,
+        });
+      }, /Provided types must match the value type passed to `params`/);
+      done();
+    });
+
+    it('should throw for invalid provided types with positional params', done => {
+      const INVALID_TYPES = ['invalid'];
+
+      assert.throws(() => {
+        bq.createQueryJob({
+          query: QUERY_STRING,
+          params: POSITIONAL_PARAMS,
+          types: INVALID_TYPES,
+        });
+      }, /Invalid type provided./);
+      done();
+    });
+
+    it('should throw for invalid type structure provided for positional params', done => {
+      assert.throws(() => {
+        bq.createQueryJob({
+          query: QUERY_STRING,
+          params: POSITIONAL_PARAMS,
+          types: NAMED_TYPES,
+        });
+      }, /Provided types must match the value type passed to `params`/);
+      done();
+    });
+
+    it('should throw for incorrect number of types provided for positional params', done => {
+      const ADDITIONAL_TYPES = ['string', 'string'];
+      assert.throws(() => {
+        bq.createQueryJob({
+          query: QUERY_STRING,
+          params: POSITIONAL_PARAMS,
+          types: ADDITIONAL_TYPES,
+        });
+      }, /Incorrect number of parameter types provided./);
+      done();
+    });
+
+    it('should throw for empty array without provided types', done => {
+      bq.createJob = (reqOpts: JobOptions) => {
+        done();
+      };
+
+      assert.throws(() => {
+        bq.createQueryJob({
+          query: QUERY_STRING,
+          params: [],
+        });
+      }, /Type must be provided for empty array./);
+      done();
+    });
+
+    it('should throw for null value without provided types', done => {
```

Comment:
Since this is already tested in the `getTypeDescriptorFromValue_` describe block, we don't need to duplicate here.